File: public/teien/world.js
teien.World = function World(app_klass) {
    var that = this;
    this.physics = new teien.Physics(this);
    this.app = new app_klass(this);
    
    this.updatePeriod = 1.0 / 60.0; // sec
    this.actors = {};
};

teien.World.prototype.setup = function(ws_uri) {
    this.ws_uri = ws_uri;
    this.ws = new WebSocket(ws_uri);
    this.ws.onmessage = this.receiveMessage.bind(this);
    
    this.physics.setup();
    
    if (typeof(this.app.setup) == 'function')
        this.app.setup();

	this.lastUpdate = 0;
	//setInterval(this.intervalHandler.bind(this), this.updatePeriod * 1000.0);
    requestAnimationFrame(this.intervalHandler.bind(this));
};

teien.World.prototype.intervalHandler = function() {
    requestAnimationFrame(this.intervalHandler.bind(this));
    
    if (this.lastUpdate == 0){
        this.lastUpdate = Date.now();
        return;
    }
        
    var now = Date.now();
    var delta = now - this.lastUpdate;
    this.lastUpdate = now;
    this.update(delta / 1000.0);  // sec

    //console.log("delta: " + delta);
};

teien.World.prototype.receiveMessage = function(ws_msg){
    var event = JSON.parse(ws_msg.data);
    switch(event.type){
    case "client_connected":
        this.clientConnected(event);
        break;
    case "client_disconnected":
        this.clientDisconnected(event);
        break;
    case "connected":
        for (var i = 0; i < event.data.length; i++) {
            //console.log("actor: " + event.data[i].name + "\n");
            this.addActor(event.data[i]);
        }
        
        if (typeof(this.app.connected) == 'function')
            this.app.connected();
        
        this.ws.send(JSON.stringify({type: 'ready'}));
        break;        
    case "actors":
        for (var i = 0; i < event.data.length; i++) {
            //console.log("actor: " + event.data[i].name + "\n");
            if (!this.hasActor(event.data[i].name)){
                this.addActor(event.data[i]);
            }
            else {
                this.actors[event.data[i].name].fromHash(event.data[i]);
            }
        }
        break;
    default:
        //console.log("No such an event: " + event.type + "\n");        
    }
};


teien.World.prototype.hasActor = function(name){
    if (this.actors[name] == undefined)
        return false;
    else
        return true;
};

teien.World.prototype.createActor = function(actor_param){
    var actor = null;
    switch(actor_param['type']){
    case 'Sphere':
        actor = new SphereActor(actor_param);
        break;
    case 'Box':
        actor = new BoxActor(actor_param);
        break;
    case 'Ghost':
        actor = new GhostActor(actor_param);
        break;                
    default:
        console.log("No such actor type: " + actor_param['type'] + "\n");
    }
    return actor;
};

teien.World.prototype.addActor = function(actor_param){
    var actor = this.createActor(actor_param);
    if (actor){
        actor.setup(this.physics);
        this.actors[actor_param.name] = actor;
        console.log("addActor: " + actor_param.name + "\n");
//        if (typeof(this.app.addActor) == 'function')
//            this.app.addActor(actor);            
    }
    return true;
}


teien.World.prototype.clientConnected = function(event) {
    console.log("A client is connected: " + event.total_clients + "\n");
};

teien.World.prototype.clientDisconnected = function(event) {
    console.log("A client is disconnected: " + event.total_clients + "\n");
};

teien.World.prototype.update = function(delta) {
    //console.log("World::update is called: " + delta + "\n");

    // Physics update
    this.physics.update(delta);
    for (var i in this.actors){
        this.actors[i].updateTransform(delta);
    }
    
    if (typeof(this.app.update) == 'function')
        this.app.update(delta);    
};

File: lib/teien/world.rb
require 'json'
require_relative './physics'

module Teien
  class World
    SYNC_PERIOD = 1.0
    attr_accessor :env    
    attr_accessor :clients
    attr_accessor :actors

    def initialize(app_klass)
      @update_period = 0
      @update_timer_thread = nil

      @env = {
        :world_aabb_min => {
          :x => -3000,
          :y =>-500,
          :z => -3000
        },
        :world_aabb_max => {
          :x => 3000,
          :y => 500,
          :z => 3000
        },
        :max_proxies => 1024 * 4,
        :gravity => {
          :x => 0,
          :y => -9.8,
          :z => 0
        }
      }

      # network  {:ws, :state}
      @clients = []

      # actor
      @actors = {}

      # physics
      @physics = Teien::Physics.new(self)

      @app = app_klass.new(self)
    end

    def add_actor(actor)
      if @actors[actor.name]
        Teien::log.info("There is already an actor which has the name[#{actor.name}].")
        return false
      else
        actor.setup(@physics)
        @actors[actor.name] = actor
        return true
      end
    end

    def send_connected(ws)
      actors = []
      @actors.each{|k, v|
        actors << v.to_hash
      }
      
      actors_event = {}
      actors_event[:type] = "connected"
      actors_event[:data] = actors

      #puts actors_event.to_json
      ws.send(actors_event.to_json)      
    end
    
    def send_actors(ws)
      actors = []
      @actors.each{|k, v|
        actors << v.to_hash
      }
      
      actors_event = {}
      actors_event[:type] = "actors"
      actors_event[:data] = actors

      #puts actors_event.to_json
      ws.send(actors_event.to_json)
    end

    #
    # events
    #

    def setup()
      @physics.setup(@env)
      @app.setup if @app.respond_to?(:setup)
      @sync_left_time = SYNC_PERIOD      
      set_update_period(1.0 / 60.0)
    end

    def connected(ws, event)
      @clients << {:ws => ws, :state => "CONNECTING"}
      Teien::log.debug("Connection's num: #{@clients.size}")
      @app.connected(ws, event) if @app.respond_to?(:connected)
      @clients.each { |client| 
        event = {:type => "client_connected", :total_clients => @clients.size}
        client[:ws].send (event.to_json)
      }

      send_connected(ws)
    end

    def disconnected(ws, event)
      @clients.each { |client|
        if (client[:ws] == ws)
          @clients.delete(client)
          break;
        end
      }
      Teien::log.debug("Connection's num: #{@clients.size}")

      @app.disconnected(ws, event) if @app.respond_to?(:disconnected)
      @clients.each { |client| 
        event = {:type => "client_disconnected", :total_clients => @clients.size}
        client[:ws].send (event.to_json)
      }

      ws = nil
    end

    def receive_message(ws, event)
      data = JSON.parse(event.data)
      case data["type"]
      when "ready" then
        @clients.each { |client|
          if (client[:ws] == ws)
            client[:state] = "READY"
            break
          end
        }
      else
        puts "receive #{data[:type]}"
      end
      
      @app.receive_message(ws, event) if @app.respond_to?(:receive_message)
    end

    def update (delta)
      #puts "delta: #{delta}"
      @physics.update(delta)
      @app.update(delta) if @app.respond_to?(:update)

      @sync_left_time -= delta
      if @sync_left_time < 0
        @clients.each { |client|
          send_actors(client[:ws]) if (client[:state] == "READY")
        }
        @sync_left_time = SYNC_PERIOD
      end
    end
    
    def run()
      Teien::log.info("Starting up a teien server.")
      setup()
    end

    private

    def set_update_period(period)
      @update_period = period

      if @update_timer_thread
        @update_timer_thread.exit
        @update_timer_thread = nil
      end

      return if @update_period <= 0

      @update_timer_thread = Thread.new {
        last_time = 0
        loop do
          start_time = Time.now
          update(start_time - last_time) unless last_time == 0
          last_time = start_time
          end_time = Time.now
          elapsed = end_time - start_time
          sleep([@update_period - elapsed, 0].max)
        end
      }
    end
  end
end
